### PR TITLE
Simplify usage of Mimir's usage of Queryables in querier

### PR DESCRIPTION
**Pull request by pstibrany**
#### What this PR does

This is internal refactoring PR. It removes support for querying multiple store queryables (not needed anymore after chunks storage was removed), and it removes `querier.QueryableWithFilter` and replaces it with `querier.Queryable`, which is similar to `storage.Queryable` but can return `nil` querier.

This is a follow-up to discussion at https://github.com/grafana/mimir/pull/4287#discussion_r1132481888

#### Checklist

- [x] Tests updated
- [na] Documentation added
- [na] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/querier/querier.go [hunk 1]
```diff
@@ -151,17 +145,15 @@ func (q *chunkQuerier) Select(sortSeries bool, hints *storage.SelectHints, match
...
 }
 
-// QueryableWithFilter extends Queryable interface with `UseQueryable` filtering function.
-type QueryableWithFilter interface {
-	storage.Queryable
-
-	// UseQueryable returns true if this queryable should be used to satisfy the query for given time range.
-	// Query min and max time are in milliseconds since epoch.
-	UseQueryable(now time.Time, queryMinT, queryMaxT int64) bool
+// Queryable interface is similar to storage.Queryable, but it can return nil querier, if query should not use this queryable.
```
Comment (56quarters) [inline pkg/querier/querier.go:148]: Can you add some extra information that would help people using this interface like:
* `Querier` will be nil if this `Queryable` shouldn't be used for a particular query. Will error be `nil` in this case?
* Is it OK for some implementations to always return a `Querier`? Never return a `Querier`?

**Review by 56quarters — APPROVED** [review]
LGTM with some minor (non-blocking) feedback.

File: pkg/querier/querier.go [hunk 2]
```diff
@@ -456,53 +444,44 @@ func (s *sliceSeriesSet) Warnings() storage.Warnings {
...
-func UseAlwaysQueryable(q storage.Queryable) QueryableWithFilter {
-	return alwaysTrueFilterQueryable{Queryable: q}
+	// Include this store only if mint is within QueryStoreAfter w.r.t current time.
+	if s.queryStoreAfter != 0 && queryMinT > util.TimeToMillis(now.Add(-s.queryStoreAfter)) {
+		return nil, nil
+	}
+	return s.q.OptionalQuerier(ctx, now, queryMinT, queryMaxT)
 }
 
 type useBeforeTimestampQueryable struct {
```
Comment (dimitarvdimitrov) [inline pkg/querier/querier.go:463]: Is this struct still needed? I can only see it being used in a single test which tests it.

**Review by dimitarvdimitrov — APPROVED** [review]
I think there's a struct we can remove. I do prefer the nil-safety of returning a `storage.NoopQuerier()` instead of nil, but that may be bike-shedding. Also agree with Nick's comment on the Queryable interface.

LGTM!

**Comment by pstibrany** [comment]
Thanks for reviews @56quarters and @dimitarvdimitrov. I also agree with you about `querier.Queryable` naming, but don't have a better idea. Do you have some suggestions?

**Comment by dimitarvdimitrov** [comment]
how about `OptionalQueryable`? `MaybeQueryable` (although this fits better as name of the method instead of the interface)?

**Review by pracucci — APPROVED** [review]
LGTM! Agree on renaming the interface to `OptionalQueryable` and remove `useBeforeTimestampQueryable`. I would merge this PR if https://github.com/grafana/mimir/pull/4287 is stuck or would benefit it.

**Comment by lamida** [comment]
The CHANGELOG has just been cut to prepare for the next Mimir release. Please rebase `main` and eventually move the CHANGELOG entry added / updated in this PR to the top of the CHANGELOG document. Thanks!

**Comment by pstibrany** [comment]
I’m looking at this with some fresh eyes, and in light of Dimitar's comment about returning `storage.NoopQuerier()` instead of `nil`, what if we simply filter out all such Noop queriers in ? Then we don’t need a new `OptionalQueryable` interface at all. 

Previously we passed consistent query time (as `now` parameter) in `UseQueryable` method, now we pass it via `Context`. It is not nice, but allows us to use upstream `storage.Queryable` without modifications.

See my latest version of the PR.

**Comment by pstibrany** [comment]
I will close this, and start again. Too many conflicts now :(

**Pull request closed without merging**